cmd/download_test.go

```diff
@@ -77,6 +77,71 @@ func TestDownloadWithoutFlags(t *testing.T) {
 	}
 }
 
+func TestSolutionFile(t *testing.T) {
+	testCases := []struct {
+		name         string
+		file         string
+		expectedPath string
+		expectedURL  string
+	}{
+		{
+			name:         "filename with special character",
+			file:         "special-char-filename#.txt",
+			expectedPath: "special-char-filename#.txt",
+			expectedURL:  "special-char-filename%23.txt",
+		},
+		{
+			name:         "filename with leading slash",
+			file:         "/with-leading-slash.txt",
+			expectedPath: "/with-leading-slash.txt",
+			expectedURL:  "/with-leading-slash.txt",
+		},
+		{
+			name:         "filename with leading backslash",
+			file:         "\\with-leading-backslash.txt",
+			expectedPath: "/with-leading-backslash.txt",
+			expectedURL:  "%5Cwith-leading-backslash.txt",
+		},
+		{
+			name:         "filename with backslashes in path",
+			file:         "\\with\\backslashes\\in\\path.txt",
+			expectedPath: "/with/backslashes/in/path.txt",
+			expectedURL:  "%5Cwith%5Cbackslashes%5Cin%5Cpath.txt",
+		},
+		{
+			name:         "path with a numeric suffix",
+			file:         "/full/path/with/numeric-suffix/bogus-track/bogus-exercise-12345/subdir/numeric.txt",
+			expectedPath: "/full/path/with/numeric-suffix/bogus-track/bogus-exercise-12345/subdir/numeric.txt",
+			expectedURL:  "/full/path/with/numeric-suffix/bogus-track/bogus-exercise-12345/subdir/numeric.txt",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			sf := solutionFile{
+				path:    tc.file,
+				baseURL: "http://www.example.com/",
+			}
+
+			expectedPath := filepath.FromSlash(tc.expectedPath)
+			if sf.relativePath() != expectedPath {
+				t.Fatalf("Expected path '%s', got '%s'", expectedPath, sf.relativePath())
+			}
+
+			url, err := sf.url()
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			expectedURL := sf.baseURL + tc.expectedURL
```

Comment:
Ah. Hm. I notice now that `tc.expectedURL` isn't really a URL, but rather a path. But we already have a `tc.expectedPath`.

I'm not sure what the best way to clarify this so we're not confused in 3 months when we've forgotten the genesis of this. I can think of a couple possible approaches that would keep us accurate:

1. ensure that `tc.expectedURL` is actually the URL, by defining it with the baseURL already in there
2. rename the fields to be `tc.expectedFilePath` and `tc.expectedURLPath` to reflect what they actually are right now

... or something else that I can't think of.